truffle/src/main/java/org/jruby/truffle/language/globals/GlobalVariableStorage.java

```diff
@@ -10,10 +10,62 @@
 
 package org.jruby.truffle.language.globals;
 
+import com.oracle.truffle.api.Assumption;
+import com.oracle.truffle.api.CompilerDirectives;
+import com.oracle.truffle.api.CompilerDirectives.CompilationFinal;
+import com.oracle.truffle.api.utilities.CyclicAssumption;
+import com.oracle.truffle.api.utilities.NeverValidAssumption;
+import org.jruby.util.cli.Options;
+
 public class GlobalVariableStorage {
-    public volatile Object value;
+
+    private static final int GLOBAL_VARIABLE_CACHE = Options.TRUFFLE_GLOBAL_VARIABLE_CACHE.load();
+
+    private final CyclicAssumption unchangedAssumption = new CyclicAssumption("global variable unchanged");
+    private int changes = 0;
+    @CompilationFinal private volatile boolean assumeConstant = true;
```

Comment:
This really means `@CompilationFinal` for compilation and `volatile` in interpreter.